File: packages/playwright/bundle.js
/**
 * ViewGraph Bundle Builder for Playwright
 *
 * Reads ViewGraph extension source modules and produces a single
 * injectable script string for Playwright's page.addScriptTag().
 * Reuses the same pattern proven in the bulk capture experiment.
 *
 * The extension modules use ES module syntax (import/export) which
 * can't run inside page.evaluate(). This bundler strips the module
 * syntax and exposes all needed functions on window.__vg.
 */

import { readFile } from 'fs/promises';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));

/** Path to the extension lib directory relative to this package. */
const EXT_LIB = join(__dirname, '..', '..', 'extension', 'lib');

/**
 * Files to bundle, in dependency order.
 * Each file's exports are collected and exposed on window.__vg.
 */
const MODULE_FILES = [
  'visibility-collector.js',
  'traverser.js',
  'salience.js',
  'serializer.js',
  'html-snapshot.js',
];

/**
 * Strip ES module import/export syntax from source code.
 * @param {string} source - Raw module source
 * @returns {string} Browser-safe source
 */
function stripModuleSyntax(source) {
  return source
    .replace(/^import\s+.*?from\s+['"].*?['"];?\s*$/gm, '')
    .replace(/^import\s+['"].*?['"];?\s*$/gm, '')
    .replace(/^export\s+(function|const|let|var|class)\s/gm, '$1 ')
    .replace(/^export\s+default\s+/gm, '')
    .replace(/^export\s*\{[^}]*\};?\s*$/gm, '');
}

/** Cached bundle string - built once per process. */
let _cached = null;

/**
 * Build the injectable script string from extension modules.
 * When running from the repo, reads source files directly.
 * When installed from npm, falls back to the pre-built bundle.
 * Cached after first call.
 * @returns {Promise<string>}
 */
export async function buildBundle() {
  if (_cached) return _cached;

  // Try reading from extension source files (repo context)
  try {
    const parts = ['(function() {', '"use strict";'];
    for (const file of MODULE_FILES) {
      const source = await readFile(join(EXT_LIB, file), 'utf-8');
      parts.push(`\n// --- ${file} ---`);
      parts.push(stripModuleSyntax(source));
    }
    parts.push('\nwindow.__vg = { checkRendered, traverseDOM, scoreElement, classifyTier, scoreAll, serialize, captureSnapshot };');
    parts.push('})();');
    _cached = parts.join('\n');
    return _cached;
  } catch {
    // Extension sources not available (npm install context) - use pre-built
  }

  // Fall back to pre-built bundle
  try {
    const { PREBUILT_BUNDLE } = await import('./bundle-prebuilt.js');
    _cached = PREBUILT_BUNDLE;
    return _cached;
  } catch {
    throw new Error(
      '@viewgraph/playwright: bundle not found. If running from the repo, ensure extension/lib/ exists. ' +
      'If installed from npm, the package may be corrupted - try reinstalling.',
    );
  }
}

File: packages/playwright-python/src/viewgraph_playwright/bundle.py
"""
Bundle loader for the ViewGraph JS capture script.

Loads the pre-built JavaScript bundle that gets injected into browser
pages via page.add_script_tag(). The bundle exposes window.__vg with
traverseDOM, scoreAll, serialize, and captureSnapshot functions.

Two loading paths:
1. Repo context: reads from packages/playwright/bundle-prebuilt.js
2. PyPI install: reads _bundle.js shipped as package data

The bundle is cached after first load (one file read per process).
"""

from __future__ import annotations

from pathlib import Path

_cached_bundle: str | None = None

# Repo-relative path to the TS package's pre-built bundle
_REPO_BUNDLE = Path(__file__).parent.parent.parent.parent / "playwright" / "bundle-prebuilt.js"

# Package-data fallback (shipped in wheel)
_LOCAL_BUNDLE = Path(__file__).parent / "_bundle.js"


def load_bundle() -> str:
    """
    Load the ViewGraph JS bundle string.

    Tries repo path first (development), then local package data (PyPI install).
    Raises FileNotFoundError if neither exists.

    Returns:
        The JavaScript IIFE string ready for page.add_script_tag().
    """
    global _cached_bundle
    if _cached_bundle is not None:
        return _cached_bundle

    # Try repo context first
    if _REPO_BUNDLE.exists():
        content = _REPO_BUNDLE.read_text(encoding="utf-8")
        # Extract the IIFE string from the ESM export
        _cached_bundle = _extract_bundle_string(content)
        return _cached_bundle

    # Fall back to local package data
    if _LOCAL_BUNDLE.exists():
        _cached_bundle = _LOCAL_BUNDLE.read_text(encoding="utf-8")
        return _cached_bundle

    raise FileNotFoundError(
        "ViewGraph JS bundle not found. "
        "If running from the repo, ensure packages/playwright/bundle-prebuilt.js exists "
        "(run 'node packages/playwright/scripts/build-bundle.js'). "
        "If installed from PyPI, the package may be corrupted - try reinstalling."
    )


def _extract_bundle_string(content: str) -> str:
    """
    Extract the IIFE string from bundle-prebuilt.js ESM export.

    The file contains: export const PREBUILT_BUNDLE = `(function() { ... })();`;
    We need just the IIFE string without the ESM wrapper.
    """
    # Find the template literal content between backticks
    start = content.find("`")
    end = content.rfind("`")
    if start == -1 or end == -1 or start == end:
        # Not a template literal - return as-is (might be raw IIFE)
        return content.strip()
    return content[start + 1 : end]
